apiserver/internal/handlers/resource/resourceunit.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resource
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v6"
+
+	"github.com/juju/juju/core/logger"
+	coreresource "github.com/juju/juju/core/resource"
+	"github.com/juju/juju/rpc/params"
+)
+
+type ResourceOpenerGetter interface {
+	Opener(*http.Request, ...string) (coreresource.Opener, error)
+}
+
+// UnitResourcesHandler is the HTTP handler for unit agent downloads of
+// resources.
+type UnitResourcesHandler struct {
+	logger               logger.Logger
+	resourceOpenerGetter ResourceOpenerGetter
+}
+
+// NewUnitResourcesHandler returns a new HTTP handler for unit agent downloads
+// of resources.
+func NewUnitResourcesHandler(
+	logger logger.Logger,
+	resourceOpenerGetter ResourceOpenerGetter,
+) *UnitResourcesHandler {
+	return &UnitResourcesHandler{
+		logger:               logger,
+		resourceOpenerGetter: resourceOpenerGetter,
+	}
+}
+
+// ServeHTTP implements http.Handler.
+func (h *UnitResourcesHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	switch req.Method {
+	case "GET":
+		if err := h.serveGet(resp, req); err != nil {
+			if err := sendError(resp, err, h.logger); err != nil {
+				h.logger.Errorf("%v", err)
+			}
+		}
+	default:
+		if err := sendError(
+			resp,
+			errors.MethodNotAllowedf("unsupported method: %q", req.Method),
+			h.logger,
+		); err != nil {
+			h.logger.Errorf("%v", err)
+		}
+	}
+}
+
+func (h *UnitResourcesHandler) serveGet(resp http.ResponseWriter, req *http.Request) error {
+	opener, err := h.resourceOpenerGetter.Opener(req, names.UnitTagKind, names.ApplicationTagKind)
+	if err != nil {
+		return err
+	}
+
+	name := req.URL.Query().Get(":resource")
```

Comment:
Show we validate the name is empty before even doing any more work, or does the Opener take care of that?

Reply:
It will go through a few layers but it'll get caught in the service layer, the first thing the opener does once it has the lock is call through to that. IMO its not worth the extra logic.